0035-ilp-over-http/0035-ilp-over-http.md

```diff
@@ -56,3 +52,123 @@ All ILP Packets MUST be returned with the HTTP status code `200: OK`.
+ 
+This extra header allows for the identity of the authentication request to be separated from authentication token itself, which reduces computational overhead as well as data-management complexity (e.g., implementations do not need to create data-store indexes using derivations of tokens for lookup purposes).
+
+Depending on the implementation, the value of this header SHOULD represent an `ILP Account Id`, but future implementations MAY make this represent an `ILP Peer Id` or some other value.
+
+##### Example Usage
+An example shared-secret in this profile is `HEiMCp0FoAC903QHueY89gAWJHo/izaBnJU8/58rlSI=`. This shared secret is passed as an `Authorization` header in each HTTP request, using the Bearer token scheme, along with an `Auth-Principal` header like this: 
+
+```
+Auth-Principal: alice-usd-123
+Authorization: Bearer HEiMCp0FoAC903QHueY89gAWJHo/izaBnJU8/58rlSI=
+``` 
+
+Implementations MAY support this profile, but SHOULD consider it for development purposes only.
+ 
+##### Trade-off Summary
+* **Pros**
+  * The simplest, most usable Authentication profile -- just a shared-secret with _at least_ 32 bytes and an identity header.
+  * Very little processing time to verify a token (just a simple value comparison or, for more security, an HMAC-SHA256 computation).
+
+* **Cons**
+  * The shared-secret is transmitted "on the wire" for every request, increasing the chances that it might be intercepted by a compromised TLS session (e.g., a [MITM attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack)); a TLS termination endpoint (e.g., a Load Balancer); or logged by an internal system during transit.
+  * The shared-secret itself never expires, so if an implementation neglects to rotate the secret with its peer, this token will likely be very long-lived. This increases the chance of compromise by an attacker, and means compromised usage of this type of token could go undetected for very longer periods of time.
+  * Requires out-of-band communication for both peers to agree upon the shared secret.
+
+#### `JWT_HS_256` Authentication Profile
+This profile allows two ILP nodes to utilize a previously agreed-upon shared-secret, but then derive an [RFC-7519](https://tools.ietf.org/html/rfc7519) compliant JWT token in order to perform actual authentication.
+
+##### JWT Claims
+In order to be considered a valid JWT for this profile, the signed JWT MUST contain a `sub` (subject) claim containing the identifier of the "principal" that the token authenticates. Depending on the implementation, the value of this claim SHOULD represent an `ILP Account Id`, but future implementations MAY make this represent an `ILP Peer Id` or some other value.
+
+A JWT token SHOULD also include an `exp` (expiry) claim that indicates a date/time after which the token should be considered invalid. Implementations SHOULD reject any tokens with a missing or invalid expiry claim. 
```

Comment:
In JWT-land, there are two _other_ **very common** claims that are typically included in a JWT, and they are probably worth talking about in the context of ILP-over-HTTP:

* The [`iss` claim](https://tools.ietf.org/html/rfc7519#section-4.1.1) identifies the principal that issued a JWT (in ILP-over-HTTP, this would be the URL of the sender of an ILP Prepare packet).
* The [`aud` claim](https://tools.ietf.org/html/rfc7519#section-4.1.3) narrows the scope of the token to only being valid at a particular peer (in ILP-over-HTTP, this would be the URL of the receiver of an ILP Prepare packet).

RFC-7519 clearly marks these two claims as optional, though it does define concrete behaviors _if_ these claims are present in a JWT.

For the purposes of this Authentication Profile in this RFC (i.e., `JWT_HS_256`), I can essentially think of two attack vectors where these claims _may_ have an impact:

* **Shared Secret is Compromised**: If an attacker were to compromise the underlying shared-secret used to generate the `HS_256` JWTs, then the attacker could simply forge the `iss` and `aud` claims, so including them in tokens conforming to this RFC doesn't seem to add any protection against that scenario.

* **Token Compromised**: If an attacker were to somehow intercept a JWT in-transit (or in a log file or wherever), the attacker would not be able to mint _new_ tokens, but the attacker could make calls against either Peer in an ILP-over-HTTP relationship. Adding the `aud` claim would help to narrow the places where this token could work to _only one of the peers_. This is because during JWT verification, each peer would verify the `aud` claim, and without the underlying shared-secret, this value could not be forged. For example, if my URL is `https://david.example.com`, and I receive a token with an `aud` of `https://alice.example.com`, I would reject this token because it was likely generated by me and meant for my peer (`alice` in this case). In other words, I would only accept tokens containing a signed `aud` claim with a value of `https://david.example.com`.

I'm inclined to update this proposal to include the `aud` claim as a normative requirement for two reasons:

1. It reduces the attack space between any two ILP-over-HTTP peers by 50% in the event of a compromised token.
1. Both Java libraries that I have investigated require the `iss` and `aud` claims, so I would like to confirm if there might be something deeper at play from a JWT security perspective relating to these two claims. I have an [outstanding question in one of the projects](https://github.com/auth0/java-jwt/issues/334) to see if that's true.